packages/contracts/src/contracts/anchor.ts

```diff
@@ -110,26 +107,46 @@ export class AnchorContract {
       .sort((a, b) => a.args.leafIndex - b.args.leafIndex) // Sort events in chronological order
       .map((e) => e.args.commitment);
 
-    const commitments = [...storedContractInfo.leaves, ...newCommitments];
-
-    // extract the commitments from the events, and update the storage
-    await this.mixersInfo.setMixerStorage(this._contract.address, currentBlock, commitments);
-
-    return commitments;
+    return {
+      lastQueriedBlock: currentBlock,
+      newLeaves: newCommitments,
+    };
   }
 
   /*
    * Generate Merkle Proof
    *  This will
-   *  1- Fetch leaves
-   *  2- build the merkle tree & get the commitment leaf index
-   * 	3- Get the Merkle Path
+   *  1- Create the merkle tree with the leaves in local storage
+   *  2- Fetch the missing leaves
+   *  3- Insert the missing leaves
+   *  4- Compare against historical roots before adding to local storage
+   *  5- return the path to the leaf.
    * */
   async generateMerkleProof(deposit: Deposit) {
-    const leaves = await this.getDepositLeaves();
-    logger.trace(`Leaves ${leaves.length}`, leaves);
-    const tree = new MerkleTree('eth', 20, leaves);
-    let leafIndex = leaves.findIndex((commitment) => commitment == bufferToFixed(deposit.commitment));
+    const storedContractInfo = await this.mixersInfo.getMixerStorage(this._contract.address);
+    const tree = new MerkleTree('eth', 20, storedContractInfo.leaves);
+
+    // Query for missing blocks starting from the stored endingBlock
+    const lastQueriedBlock = storedContractInfo.lastQueriedBlock;
+
+    const fetchedLeaves = await this.getDepositLeaves(lastQueriedBlock + 1);
+    logger.trace(`New Leaves ${fetchedLeaves.newLeaves.length}`, fetchedLeaves.newLeaves);
+
+    tree.batch_insert(fetchedLeaves.newLeaves);
+    const newRoot = tree.get_root();
+    const formattedRoot = bufferToFixed(newRoot);
+
+    // compare root against contract, and store if there is a match
+    if (this._contract.isKnownRoot(formattedRoot)) {
+      this.mixersInfo.setMixerStorage(this._contract.address, lastQueriedBlock, [
+        ...storedContractInfo.leaves,
+        ...fetchedLeaves.newLeaves,
+      ]);
+    }
+
+    let leafIndex = [...storedContractInfo.leaves, ...fetchedLeaves.newLeaves].findIndex(
+      (commitment) => commitment == bufferToFixed(deposit.commitment)
```

Comment:
@AhmedKorim @nepoche since the user is providing a deposit to withdraw, I would suggest we find a way for the user to also possess their deposit leaf's index a priori. This would allow us to remove this call entirely which may improve speed as we grow and at least simplify some things.

For example, as we work through this more, consider all the ways in which the user can provide information in place of us querying it. The protocol is guaranteed to work if the user provides the right inputs, therefore if they mess up or provide the wrong inputs it will simply fail.. the burden becomes more on the user.

Reply:
If it is a piece of data not completely controlled via the client, I think we could run into some issues with transaction ordering.